executor/executor_test.go

```diff
@@ -5357,6 +5358,122 @@ func TestHistoryRead(t *testing.T) {
 	tk.MustQuery("select * from history_read order by a").Check(testkit.Rows("2 <nil>", "4 <nil>", "8 8", "9 9"))
 }
 
+func TestHistoryReadInTxn(t *testing.T) {
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+
+	// For mocktikv, safe point is not initialized, we manually insert it for snapshot to use.
+	safePointName := "tikv_gc_safe_point"
+	safePointValue := "20060102-15:04:05 -0700"
+	safePointComment := "All versions after safe point can be accessed. (DO NOT EDIT)"
+	updateSafePoint := fmt.Sprintf(`INSERT INTO mysql.tidb VALUES ('%[1]s', '%[2]s', '%[3]s')
+    ON DUPLICATE KEY
+    UPDATE variable_value = '%[2]s', comment = '%[3]s'`, safePointName, safePointValue, safePointComment)
+	tk.MustExec(updateSafePoint)
+
+	tk.MustExec("drop table if exists his_t0, his_t1")
+	tk.MustExec("create table his_t0(id int primary key, v int)")
+	tk.MustExec("insert into his_t0 values(1, 10)")
+
+	time.Sleep(time.Millisecond)
+	tk.MustExec("set @a=now(6)")
+	time.Sleep(time.Millisecond)
+	tk.MustExec("create table his_t1(id int primary key, v int)")
+	tk.MustExec("update his_t0 set v=v+1")
+	time.Sleep(time.Millisecond)
+	tk.MustExec("set tidb_snapshot=now(6)")
+	ts2 := tk.Session().GetSessionVars().SnapshotTS
+	tk.MustExec("set tidb_snapshot=''")
+	time.Sleep(time.Millisecond)
+	tk.MustExec("update his_t0 set v=v+1")
+	tk.MustExec("insert into his_t1 values(10, 100)")
+
+	init := func(isolation string, setSnapshotBeforeTxn bool) {
+		if isolation == "none" {
+			tk.MustExec("set @@tidb_snapshot=@a")
+			return
+		}
+
+		if setSnapshotBeforeTxn {
+			tk.MustExec("set @@tidb_snapshot=@a")
+		}
+
+		if isolation == "optimistic" {
+			tk.MustExec("begin optimistic")
+		} else {
+			tk.MustExec(fmt.Sprintf("set @@tx_isolation='%s'", isolation))
+			tk.MustExec("begin pessimistic")
+		}
+
+		if !setSnapshotBeforeTxn {
+			tk.MustExec("set @@tidb_snapshot=@a")
+		}
+	}
+
+	for _, isolation := range []string{
+		"none", // not start an explicit txn
+		"optimistic",
+		"REPEATABLE-READ",
+		"READ-COMMITTED",
+	} {
+		for _, setSnapshotBeforeTxn := range []bool{false, true} {
+			t.Run(fmt.Sprintf("[%s] setSnapshotBeforeTxn[%v]", isolation, setSnapshotBeforeTxn), func(t *testing.T) {
+				tk.MustExec("rollback")
+				tk.MustExec("set @@tidb_snapshot=''")
+
+				init(isolation, setSnapshotBeforeTxn)
+				// When tidb_snapshot is set, should use the snapshot info schema
+				tk.MustQuery("show tables like 'his_%'").Check(testkit.Rows("his_t0"))
+
+				// When tidb_snapshot is set, select should use select ts
+				tk.MustQuery("select * from his_t0").Check(testkit.Rows("1 10"))
+				tk.MustQuery("select * from his_t0 where id=1").Check(testkit.Rows("1 10"))
+
+				// When tidb_snapshot is set, select ... for update should not be effected
```

Comment:
Does here mean `should be affected`? It's also needed to cover the `update/delete` paths which may do for update read first. 

The expectations seem to be not clear when `tidb_snapshot` is used together with pessimistic transaction semantics. /cc @MyonKeminta @you06

Reply:
I think it can be "select ... for update should not be allowed"